Source/WebCore/page/LocalFrameView.cpp

```diff
@@ -1321,7 +1321,139 @@ void LocalFrameView::updateScrollGeometryContentSize()
     if (!page || !page->chrome().client().needsScrollGeometryUpdates())
         return;
 
-    m_scrollGeometryContentSize = contentsSize();
+    RefPtr document = m_frame->document();
+    if (!document)
+        return;
+
+    IntSize documentSize = layoutSize();
+    CheckedPtr<RenderBox> documentRenderer;
+
+    IntSize bodySize;
+    CheckedPtr<RenderBox> bodyRenderer;
+
+    bool heightDeterminingBoxIsPositioned = false;
+
+    IntSize scrollGeometryContentSize;
+    auto updateScrollGeometryContentSizeWithBoxAndPosition = [&scrollGeometryContentSize, &heightDeterminingBoxIsPositioned](const RenderBox& box, IntPoint position) {
+        scrollGeometryContentSize.setWidth(std::max(scrollGeometryContentSize.width(), position.x()));
+
+        auto oldHeight = scrollGeometryContentSize.height();
+        scrollGeometryContentSize.setHeight(std::max(scrollGeometryContentSize.height(), position.y()));
+
+        if (oldHeight != scrollGeometryContentSize.height())
+            heightDeterminingBoxIsPositioned = box.isPositioned();
+    };
+
+    auto maxPositionForBox = [](const RenderBox& box) {
+        auto borderBoxRect = snappedIntRect(box.borderBoxRect());
+        auto layoutOverflowRect = snappedIntRect(box.layoutOverflowRect());
+
+        auto borderBoxRectMaxCorner = borderBoxRect.maxXMaxYCorner();
+        auto layoutOverflowRectMaxCorner = layoutOverflowRect.maxXMaxYCorner();
+
+        auto rect = borderBoxRect;
+        if (box.effectiveOverflowX() == Overflow::Visible && layoutOverflowRectMaxCorner.x() > borderBoxRectMaxCorner.x())
+            rect.setWidth(borderBoxRect.width() + layoutOverflowRectMaxCorner.x() - borderBoxRectMaxCorner.x());
+        if (box.effectiveOverflowY() == Overflow::Visible && layoutOverflowRectMaxCorner.y() > borderBoxRectMaxCorner.y())
+            rect.setHeight(borderBoxRect.height() + layoutOverflowRectMaxCorner.y() - borderBoxRectMaxCorner.y());
+
+        auto point = roundedIntPoint(box.localToAbsolute(rect.location()));
+        rect.setLocation(point);
+
+        return rect.maxXMaxYCorner();
+    };
+
+    CheckedPtr<RenderObject> renderer = renderView()->firstChild();
```

Comment:
is the template parameter explicitly needed here?

Reply:
This can probably be skipped, yes